L80-csharp-load-balancing.md

```diff
@@ -0,0 +1,387 @@
+        public ConnectivityState State { get; }
+        public Task WaitForStateChangedAsync(ConnectivityState lastObservedState, CancellationToken cancellationToken = default);
+    }
+    public enum ConnectivityState {
+        Idle = 0,
+        Connecting = 1,
+        Ready = 2,
+        TransientFailure = 3,
+        Shutdown = 4,
+    }
+}
+    
+namespace Grpc.Net.Client.Balancer {
+    public sealed class ResolverOptions {
+        public bool DisableServiceConfig { get; }
+    }
+    public sealed class SubchannelOptions {
+        public SubchannelOptions(IReadOnlyList<DnsEndPoint> addresses);
+        public IReadOnlyList<DnsEndPoint> Addresses { get; }
+    }
+    public interface IChannelControlHelper {
+        Subchannel CreateSubchannel(SubchannelOptions options);
+        void RefreshResolver();
+        void UpdateState(BalancerState state);
+    }
+    public sealed class BalancerAttributes : IDictionary<string, object?>, ICollection<KeyValuePair<string, object?>>, IEnumerable<KeyValuePair<string, object?>>, IEnumerable {
+        public BalancerAttributes();
+        public static readonly BalancerAttributes Empty;
+        int ICollection<KeyValuePair<string, object>>.Count { get; }
+        bool ICollection<KeyValuePair<string, object>>.IsReadOnly { get; }
+        object? IDictionary<string, object>.Item[string key] { get; set; }
+        ICollection<string> IDictionary<string, object>.Keys { get; }
+        ICollection<object?> IDictionary<string, object>.Values { get; }
+        public void Set<TValue>(BalancerAttributesKey<TValue> key, TValue value);
+        public bool TryGetValue<TValue>(BalancerAttributesKey<TValue> key, out TValue value);
+        void ICollection<KeyValuePair<string, object>>.Add(KeyValuePair<string, object?> item);
+        void ICollection<KeyValuePair<string, object>>.Clear();
+        bool ICollection<KeyValuePair<string, object>>.Contains(KeyValuePair<string, object?> item);
+        void ICollection<KeyValuePair<string, object>>.CopyTo(KeyValuePair<string, object?>[] array, int arrayIndex);
+        bool ICollection<KeyValuePair<string, object>>.Remove(KeyValuePair<string, object?> item);
+        void IDictionary<string, object>.Add(string key, object? value);
+        bool IDictionary<string, object>.ContainsKey(string key);
+        object? IDictionary<string, object>.get_Item(string key);
+        bool IDictionary<string, object>.Remove(string key);
+        void IDictionary<string, object>.set_Item(string key, object? value);
+        bool IDictionary<string, object>.TryGetValue(string key, out object? value);
+        IEnumerator<KeyValuePair<string, object?>> IEnumerable<KeyValuePair<string, object>>.GetEnumerator();
+        IEnumerator IEnumerable.GetEnumerator();
+    }
+    public readonly struct BalancerAttributesKey<TValue> {
+        public BalancerAttributesKey(string key);
+        public string Key { get; }
+    }
+    public sealed class BalancerState {
```

Comment:
Github won't let me add a comment on the following thread, so continuing it here:

https://github.com/grpc/proposal/pull/240#discussion_r636283136

This still needs a way to return a status when the state is TRANSIENT_FAILURE, so that LB policies can return useful error messages when they fail individual RPCs.

Reply:
That is supported. The status comes from the picker. It returns a result with a non-OK status - `PickResult.Status` - and that is reported to the user.

Reply:
Yes, that is supported. When the subchannel fails to connect to the single address then the balancer state is set to TRANSIENT_FAILURE and the picker is set to an instance that will return status with information about the connection failure.

I wrote a test that follows your example and this is the error from the RPC:

>     Grpc.Core.RpcException: Status(StatusCode="Unavailable", Detail="Error connecting to subchannel.", DebugException="System.Net.Sockets.SocketException (10061): No connection could be made because the target machine actively refused it.
>        at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
>        at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
>        at System.Net.Sockets.Socket.<ConnectAsync>g__WaitForConnectWithCancellation|277_0(AwaitableSocketAsyncEventArgs saea, ValueTask connectTask, CancellationToken cancellationToken)
>        at Grpc.Net.Client.Balancer.Internal.SocketConnectivitySubchannelTransport.TryConnectAsync(CancellationToken cancellationToken) in C:\Development\Source\grpc-dotnet\src\Grpc.Net.Client\Balancer\Internal\SocketConnectivitySubchannelTransport.cs:line 123")
>        at Grpc.Net.Client.Balancer.Internal.ConnectionManager.PickAsync(PickContext context, Boolean waitForReady, CancellationToken cancellationToken) in C:\Development\Source\grpc-dotnet\src\Grpc.Net.Client\Balancer\Internal\ConnectionManager.cs:line 379
>        at Grpc.Net.Client.Balancer.Internal.BalancerHttpHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in C:\Development\Source\grpc-dotnet\src\Grpc.Net.Client\Balancer\Internal\BalancerHttpHandler.cs:line 78
>        at Grpc.Net.Client.Internal.GrpcCall`2.RunCall(HttpRequestMessage request, Nullable`1 timeout) in C:\Development\Source\grpc-dotnet\src\Grpc.Net.Client\Internal\GrpcCall.cs:line 457

The `SocketException` is the low-level reason for the failure:

> System.Net.Sockets.SocketException (10061): No connection could be made because the target machine actively refused it.

Reply:
How does this work if you're not returning the status with the connectivity state?  Is there some other API by which the LB policy gets this status back from the subchannel?  This seems different from all of the other languages, so I'd like to understand what the reason is for this difference.

Reply:
I'm a little confused. I thought we were talking about `BalancerState` which is an aggregate state of the subchannels.

State updates for individual channels are provided via SubchannelState which has a status property:

```cs
    public sealed class SubchannelState {
        public ConnectivityState State { get; }
        public Status Status { get; }
    }
```

What would you like to see changed on `BalancerState`? Today the status is inside the `SubchannelPicker`. Would you prefer that there is an explicit `Status` property? And when `BalancerState` is created, it is either given a picker, or a non-OK status?

e.g.

```diff
    public sealed class BalancerState {
        public BalancerState(ConnectivityState connectivityState, SubchannelPicker picker);
+       public BalancerState(ConnectivityState connectivityState, Status status);
        public ConnectivityState ConnectivityState { get; }
-       public SubchannelPicker Picker { get; }
+       public SubchannelPicker? Picker { get; }
+       public Status? Status { get; }
    }
```

Reply:
I just checked with Eric and Doug, and it looks like neither Java nor Go actually have this.  So I'll let you decide whether or not to add this.

I think if you have a way to add it later without breaking existing callers, that's probably fine.

Reply:
It can be added in the future if required.